Added CSharp13 Support

dersia opened the pull request:
This PR adds `CSharp13` as another language.

This will have the following effects:
* generated CSharp code will use [File Scoped Namespaces](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/file-scoped-namespaces)
* reverts null checking back to pattern matching (was changed in https://github.com/microsoft/kiota/pull/3842)
* sets `#nullable enable` at the beginning of generated files
* removes all the pragma for `NETSTANDARD2_1_OR_GREATER || NETCOREAPP3_1_OR_GREATER`

this also fixes: https://github.com/microsoft/kiota/issues/3944

Different than the PR https://github.com/microsoft/kiota/pull/3945 this keeps backward compatibility for net framework and older C# versions.
(@0xced thanks for the PR it inspired me to this PR)

It also allows us to incorporate newer C# feature in the future when generating C# code.

dersia committed:
Added CSharp13 Support

Commit:
chore: dotnet format

andrueastman committed:
Merge branch 'main' into main

andrueastman committed:
Merge branch 'main' into main

baywet reviewed (changes requested):
Thanks for the contribution!

I don't think we'll be able to accept this pull request. Fundamentally this is because it doubles up the support cost for C#.

When we started kiota, over 4 years ago now. The dotnet landscape was really different:
- net core 3.1 to 6 were still supported https://endoflife.date/dotnet
- net fx 4.5.2, 4.6 and 4.6.1 were still supported https://endoflife.date/dotnetfx
- visual studio 2012 and 2013 were still supported https://endoflife.date/visual-studio
- close to 20% of NEW PROJECTS were still created targeting net fx

Now the situation is different:
- only net core 8 and up are supported
- only net fx 4.6.2 and up are supported
- in about 8 months only VS 2017 and up will be supported (2025 still has about 8 months of support remaining)
- about 1% of new projects are still created targeting netfx (but there's a world of existing projects being maintained)

The main reason why we (the kiota team) decided to support an older version of the language was so that large audience who bought and ran products that are still support didn't get left behind.
Now, that audience has reduced drastically, and VS2019 [supports targeting net5 and up](https://learn.microsoft.com/en-us/visualstudio/releases/2019/compatibility) so I think it'd be time to re-evaluate and potentially drop support for net framework. Which would unlock all modern language features.
To be clear, this is my personal opinion, which still needs to go through management approval, funding, etc... And when this gets approved, we'll need to major rev kiota in the process [as per our support policy](https://learn.microsoft.com/en-us/openapi/kiota/support#major-releases)

I hope this explanation shine lights on the current "status quo".
Let us know if you have any additional comments or questions.

darrelmiller commented:
@baywet What could we do to enable people like @dersia without taking on additional maintenance in the core?  
I guess one option is a fork, but that makes it difficult for community extensions to be deployed and distributed without creating a whole new parallel distribution.  And that will cause customer confusion for everyone.

We should revisit the notion of pluggable language support and see if the .net ecosystem has better options/solutions than it did 4 years ago.

I do agree though that having two C# variants in the core of Kiota is problematic from a customer adoption and support perspective.

baywet commented:
@darrelmiller regardless of the module loading approach and its ergonomic/security/discoverability considerations for the end user, this approach will also bring significant additional maintenance costs, I'd argue they are greater than "doubling up" specific languages.

We'd effectively need to:
- (set up infrastructure to discover, load and select those modules)
- maintain source compatibility on the DOM interface.
- maintain behaviour compatibility on the DOM interface.
- document the interfaces [like OpenAPI generator does today](https://openapi-generator.tech/docs/templating/#models)

In any case, [the guidance hasn't changed much](https://learn.microsoft.com/en-us/dotnet/core/tutorials/creating-app-with-plugin-support):
1. define interfaces
2. ship an assembly which implements them
3. use dynamic assembly loading and reflection to load the implementations (which doesn't work with trimming a and a few other contexts)

There seems to be [a new plugin definition approach](https://learn.microsoft.com/en-us/dotnet/core/tutorials/creating-app-with-plugin-support#simple-plugin-with-no-dependencies), but it's simply hiding away some of the complexity and fundamentally relying on the same approach.

Some open source projects [like this one](https://github.com/weikio/PluginFramework) help bootstrap the catalog discovery and inventory aspects.

dersia commented:
Hi @baywet,

I appreciate your reply. I would kindly disagree with you on the point, that it would double the support cost for c#. I implemented it exactly this way to avoid this. This implementation does exactly the same as the `Csharp` language implementation does. I just refactored a little bit of the code, some for efficiency and some for the features that I added, and then just used `if` statements in 3 places to implement the cleaner output.

It doesn't take much to keep this up, as it is filled on a bool and just avoids outputting all the necessary `#if` pragmas to support netfx. that's it.

Thinking of the future of kiota and what is coming to c# we will surly have to implement something like this. Once union support is in c# we need to change a few things in kiota. While looking through the code I realized that right now it is using refiners to resolve unions and then there are places in the code that make sure that unions are implemented in a compatible way. Once union types arrive, we will need that code for netfx and older version, but surly kiota has to support real c# unions if we want user acceptance. that will have a bigger impact than this change.

Although I have to admit, I am not very happy about the name `CSharp13` and have thought about alternatives like `CSharpNext` so it can roll forward. I also considered just adding a commandline flag but I had the issue that I couldn't come up with a name that describes "all" the changes except `--csharp-13-support`, but even then it would behave the same way.

And please don't get me wrong, I like the support for netfx, I do have some customers that are still on netfx and I know, that those will profit form such support and they will greatly appreciate the effort, however I also see customers like myself, who want to use kiota for client generation, but they also want to have cleaner, more readable and efficient code. This can only be guaranteed by splitting the implementation, just like there is a split between netfx and net (core). 

@darrelmiller thank your for your intervention, but I don't think a plugin system for languages is a good approach for kiota unless we would define a full-featured intermediate language. On the otherside I would love to see a plugin system to support other specs than just OpenApi, i.e. Typespec (without going through OpenAPI Generation), or protobuf. Sure in this scenario a intermediate language would help a lot.

With all that being said, I hope @baywet can take time to look more thoroughly through this PR and reconsider it, since it isn't really changing much and the support cost will stay small.

PS:
I hope we will still take efficiency changes and the small refactoring even if the feature itself won't be accepted.

baywet commented:
Thank you for your patience.

Although I appreciate the effort to minimal code changes, the additional maintenance is not only in the code, but also in the documentation, integration tests, etc...
It also creates a precedent for other languages, what if TS developers now want a TSNext, and then Go? etc...

Our team is simply not sufficiently staffed to maintain multiple versions of the same language.

I did spend some time looking at the content and in addition to the high level changes you've mentioned when creating the PR, there's also a small refactoring in code property writer to avoid duplicated code. This change is orthogonal and could be merged independently.

Looking at the other changes:
- File scoped namespace, [shipped with C#10](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-version-history#c-version-10), mostly esthetic, compile time feature.
- reverts null checking back to pattern matching, [shipped with C#8](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-version-history#c-version-80). As far as I understand this one is mostly esthetic and doesn't provide any functional/performance benefit, I suggest we don't even bother with that. It's not clear to me whether it's just compile time or has implications at runtime/cannot be gracefully downgraded by the compiler when needed
- sets #nullable enable at the beginning of generated files and removed conditional compilation for nullable, [also shipped with C#8](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-version-history#c-version-80), compile time feature.

As far as I can tell:

- C#10 mostly matches with .NET6 / VS2022
- C#8 mostly matches with .NET core 3.1 / VS2019


Based on this analysis and my previous comment, in 8 months from now, only VS2017 will still be supported and NOT support C#8, which is arguably the most important change in the batch from an experience point of view. If we do decide to bump the requirement to C#8, [there's still a work around](https://stackoverflow.com/questions/54701377/how-can-i-use-c-sharp-8-with-visual-studio-2017) but I suspect most people won't want that, and will be left behind.

If we wanted to "raise the bar" of an existing stable language **without** a major bump, we'd need to [amend the support policy first](https://learn.microsoft.com/en-us/openapi/kiota/support) to add an exception dropping support for product versions that are not supported anymore. This is still a concern since, C# can be written in products not owned by Microsoft, Java has multiple SDKs (Android, Oracle, Adoptium...), etc...

All of that to say, I don't think we can take in those changes, which are only esthetic, without:
1. adding significant additional maintenance burden.
2. breaking our support policy.
3. requiring a major revision, which is being planned.

Let me know if you have any additional comments or questions.

arledesma commented:
@MIchaelMainer I wanted to check in and see if there has been any changes regarding this type of language support request. 

Would love to not exclude, or fix up, kiota generated code to satisfy our SAST rules.
